_posts/2025-04-28-sparse-autodiff.md

```diff
@@ -664,5 +664,222 @@ This symmetry enables **colorings with fewer colors**, reducing the complexity o
 Moreover, symmetric coloring becomes especially advantageous when the Hessian needs to be recomputed for multiple values of $x$, as the reduced number of colors amortizes the initial expense.
 
-
 ## Demonstration
+
+We conclude this blog post with a demonstration of automatic sparse differentiation in a high-level programming language, namely the [Julia language](https://julialang.org/).
+While still at an early stage of development, we hope that such an example of unified pipeline for sparse Jacobians and Hessians can inspire developers in other languages to revisit ASD.
+
+<aside class="l-body box-note" markdown="1">
+The authors of this blog post are all developers of the ASD ecosystem in Julia. We are not aware of a similar ecosystem in Python or R, which is why we chose Julia to present it.
+</aside>
+
+### Necessary packages
+
+Here are the packages we will use for this demonstration.
+We also use a few other packages for visualization.
+
+| Package | Purpose |
+|---|---|
+| [SparseConnectivityTracer.jl](https://github.com/adrhill/SparseConnectivityTracer.jl) | Sparsity pattern detection with operator overloading. |
+| [SparseMatrixColorings.jl](https://github.com/gdalle/SparseMatrixColorings.jl) | Greedy algorithms for colorings, decompression utilities. |
+| [ForwardDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl) | Forward-mode AD and computation of JVPs. |
+| [DifferentiationInterface.jl](https://github.com/JuliaDiff/DifferentiationInterface.jl) | High-level interface bringing all of these together. |
+
+As with any language, the first step is importing the dependencies.
+
+```julia
+using DifferentiationInterface  # no prefix necessary
+import SparseConnectivityTracer  # requires prefix
+import SparseMatrixColorings
+import ForwardDiff
+```
+
+### Test function
+
+As our test function, we choose a very simple iterated difference operator.
+It takes a vector $\mathbf{x} \in \mathbb{R}^n$ and outputs a slightly shorter vector $y \in \mathbb{R}^{n-k}$ depending on the number of iterations $k$.
+In pure Julia, this is written as follows:
+
+```julia
+
+function differences(x)
+    n = length(x)
+    y = similar(x, n - 1)
+    for i in 1:(n - 1)
+        y[i] = x[i + 1] - x[i]
+    end
+    return y
+end
+
+function iterated_differences(x, k)
+    if k == 0
+        return x
+    else
+        y = iterated_differences(x, k - 1)
+        return differences(y)
+    end
+end
+```
+
+Let us check that the function returns what we expect:
+
+```julia
+julia> iterated_differences([1, 4, 9, 16], 1)
+3-element Vector{Int64}:
+ 3
+ 5
+ 7
+
+julia> iterated_differences([1, 4, 9, 16], 2)
+2-element Vector{Int64}:
+ 2
+ 2
+```
+
+### Dense Jacobian
+
+The key concept behind DifferentiationInterface.jl is that of "backend".
+There are several AD systems in Julia, with different features and tradeoff, but we access them through a common API.
+Here, we use ForwardDiff.jl as our dense backend:
+
+```julia
+dense_backend = DI.AutoForwardDiff()
```

Comment:
Why did you choose to namespace DI everywhere?
I think it makes the demo more complicated.

Reply:
Because Python users namespace everything. And since we have several packages interacting, prefixes make it clearer where each piece fits into the whole

Reply:
I think this makes our code more complicated. The point is to simplify the exposition to ASD. The simpler, the better.